Generators/include/Generators/GeneratorTParticle.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023-2099 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @author: Christian Holm Christensen <cholm@nbi.dk>
+#ifndef ALICEO2_GENERATORTPARTICLE_H_
+#define ALICEO2_GENERATORTPARTICLE_H_
+#include <FairGenerator.h>
+#include <Generators/Generator.h>
+#include <list>
+
+// Forward decls
+class TChain;
+class TParticle;
+class TClonesArray;
+
+namespace o2
+{
+namespace eventgen
+{
+/// A class that reads in particles of class @c TParticle from a
+/// branch in a @c TChain.
+///
+/// Optionally, a program that generates such a @c TTree can be
+/// spawn, and the @c TParticles written to a file from which this
+/// object reads them in.  This is done with
+///
+///  --configKeyValues "TParticle.progCmd=<eg program and options>"
+///
+/// which will execute the specified EG program with the given
+/// options.  The EG program _must_ support the command line options
+///
+///   -n NUMBER   Number of events to generate
+///   -o FILENAME Name of file to write to
+///
+/// The tree name and particle branch names can be configured.
+///
+///   --configKeyValues "TParticle.treeName=T,TParticle.branchName=P"
+///
+/// File(s) to read are also configurable
+///
+///   --configKeyValues "TParticle.fileNames=foo.root,bar.root"
+///
+class GeneratorTParticle : public Generator
```

Comment:
I am a bit unsure about this one: We already have a generator "GeneratorFromFile" which reads TParticles from a Run2 - like ROOT file.

A quick check if we still need this one would good.

Reply:
There's a difference between the `GeneratorFromFile` and `GeneratorTParticle`, mainly in that `GeneratorTParticle` 
- is more flexible 
- can read data from external program rather than from file 
- can read from a `TChain` of files 
- assumes _all_ events are in the same `TTree`, unlike `GeneratorFromFile` which assumes there's one `TTree` for each event in the `TDirectory` named `Event<_event-no_>` in the single input file. 

As such, the two classes `GeneratorFromFile` and `GeneratorTParticle serves different needs 

- `GeneratorFromFile` is designed to read from an **AliROOT** `Kinematics.root` file
- `GeneratorTParticle` is designed to read from a general **ROOT** event tree, for example produced by a `TGenerator`. 

Hope that clarifies the differences.

Reply:
Oh, and another difference.   Using the key `progCmd`, one could execute a `TGenerator` as a background process.  Say we have the script `MyEG.C`

```
class MyGenerator : public TGenerator 
{
   ...
};
 
void MyEG(const char* output, Int_t nEv, Int_t seed, Float_t bmax)
{
   gRandom->SetSeed(seed);
  
   MyGenerator generator(bMax);
  TFile* file = TFile::Open(output, "RECREATE");
  TTree* tree = new TTree("T");
  TClonesArray* particles = new TClonesArray("TParticle");
  tree->Branch("Particles", &particles);
  tree->SetDirectory(file);

  for (Int_t iEv = 0; iEv < nEv; iEv++) {
     generator.GenerateEvent();
     generator.ImportParticles(particles);
     tree->Fill();
     tree->AutoSave("SaveSelf FlushBaskets Overwrite");
  }
  file->Write();
  file->Close();
}
``` 

and a simple script like `myeg.sh` 

```
#!/bin/sh 

nev=1
out=myeg.root 
seed=0
bmax=0

while test $# -gt 0; do 
   case $1 in 
   -n|--nevents) nev=$2 ; shift ;;
   -o|--output)   out=$2 ; shift ;;
   -b|--bMax)     bmax=$2; shift;;  # Future proof 
   -s|--seed)       seed=$2; shift;;
  --help) echo "Usage: $0 [-n NEV] [-o OUTPUT] [-s SEED] [-b bmax]"; exit 0;;
  esac 
root MyEG.C++\( \"$out\",$nev,$seed,$bmax\)
```

then one can do 

```
o2-sim -g tparticle --configKeyValues "TParticle.progCmd=myeg.sh"
```

which means all the old `AliGenerators` can be used more or less directly :-)